### DISCUSSION TITLE
Hardware UART Invert

### DISCUSSION THREAD

**Discussion by carl-sauter** [earlephilhower/arduino-pico#2393] [Q&A] [answered]
Hey guys is it possible to use the UART inversion for hardware uart? I read that this isn't supported for SoftwareSerial, but nothing mentioned in the UART section.

**Comment by earlephilhower** [comment]
See #1443 .  The HW UART doesn't do inversion.

**Reply by carl-sauter** [reply]
But I can do it in MicroPython. Or is this something else?

```python
self.sbus = UART(1)
self.sbus.init(baudrate = 100000, bits=8, parity=0, stop=2, timeout_char=3, rxbuf=250, rx=pin, invert=UART.INV_RX)
```

**Reply by earlephilhower** [reply]
I don't have any info on MicroPython.  If you find a HW register they twiddle to invert, I'd be happy to pull it into the core here.  But there doesn't seem to be anything that I can find in the [datasheet](https://datasheets.raspberrypi.com/rp2350/rp2350-datasheet.pdf)...

**Reply by carl-sauter** [reply]
Maybe this helps https://talk.dallasmakerspace.org/t/raspberry-pi-pico-sbus-code-help/79375/6. The guy claims, that you can invert any gpio pin.

**Reply by earlephilhower** [reply]
Ah, did not think of that.  The peripheral won't invert, but the pads can.  I think this can easily be added to both SerialUART/PIO.  I'll throw something together over the weekend and if you could give it a test it'd be great...

**Reply by carl-sauter** [reply]
I have to admit, that I only understand parts of what you saying. Im not really into those "technical" things like assembler registers or like anything mentioned in the datasheet. But im happy to test it with my sbus receiver :)

**Reply by earlephilhower** [reply, marked as answer]
Please give #2395 a try, @Kinqdos .  I did a quick output test using a ``pulseview`` to check and it does seem to be inverting as expected with "idle" being LOW, not HIGH.

````
void setup() {
  Serial1.setInvertRX();
  Serial1.setInvertTX();
  Serial1.begin(115200);
}

int x = 0;
void loop() {
Serial1.printf("%d\n", x++);
delay(100);
}
````

![image](https://github.com/user-attachments/assets/ae9c789d-539a-4bd1-8c91-9c8f1a67861f)

**Reply by carl-sauter** [reply]
Tested it with my sbus receiver and it works fine! Thank you for the fast response and implementation!!! :)
